### PR TITLE
filterable: generalize S.justs, S.lefts, and S.rights

### PR DESCRIPTION
**Pull request by davidchambers**
<https://github.com/sanctuary-js/sanctuary/issues/489#issuecomment-366982183>

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2173,10 +2178,10 @@
   }
   S.maybe_ = def('maybe_', {}, [$.Thunk(b), Fn(a, b), $Maybe(a), b], maybe_);
 
-  //# justs :: Array (Maybe a) -> Array a
+  //# justs :: (Filterable f, Functor f) => f (Maybe a) -> f a
   //.
-  //. Takes an array of Maybes and returns an array containing each Just's
-  //. value. Equivalent to Haskell's `catMaybes` function.
+  //. Discards each of the structure's elements which is Nothing, and unwraps
```
Comment (Avaq) [inline index.js]: It was recently discussed that we could omit mentions of *the structure* from these descriptions, eg: *Discards each element which is Nothing...*; Will we do that in a separate PR?

Comment (davidchambers) [reply index.js]: I've updated the descriptions. Please have another look. :eyes:

**Comment by davidchambers** [comment]
> Will you make the changes to the signatures of derived functions in a separate PR?

I believe the only function derived from one of these functions is [`S.mapMaybe`][1]. I will open a pull request to generalize that function once this pull request has been merged. :)


[1]: https://sanctuary.js.org/#mapMaybe
